MediaBrowser.Controller/SyncPlay/Queue/PlayQueueManager.cs

```diff
@@ -0,0 +1,596 @@
+            }
+
+            _disposed = true;
+        }
+
+        /// <summary>
+        /// Gets the next available id.
+        /// </summary>
+        /// <returns>The next available id.</returns>
+        private int GetNextProgressiveId() {
+            return ProgressiveId++;
+        }
+
+        /// <summary>
+        /// Creates a list from the array of items. Each item is given an unique playlist id.
+        /// </summary>
+        /// <returns>The list of queue items.</returns>
+        private List<QueueItem> CreateQueueItemsFromArray(Guid[] items)
+        {
+            return items.ToList()
+                .Select(item => new QueueItem()
+                {
+                    ItemId = item,
+                    PlaylistItemId = "syncPlayItem" + GetNextProgressiveId()
+                })
+                .ToList();
+        }
+
+        /// <summary>
+        /// Gets the current playlist, depending on the shuffle mode.
+        /// </summary>
+        /// <returns>The playlist.</returns>
+        private List<QueueItem> GetPlaylistAsList()
+        {
+            if (ShuffleMode.Equals(GroupShuffleMode.Shuffle))
+            {
+                return ShuffledPlaylist;
+            }
+            else
+            {
+                return SortedPlaylist;
+            }
+        }
+
+        /// <summary>
+        /// Gets the current playlist as an array, depending on the shuffle mode.
+        /// </summary>
+        /// <returns>The array of items in the playlist.</returns>
+        public QueueItem[] GetPlaylist() {
+            if (ShuffleMode.Equals(GroupShuffleMode.Shuffle))
+            {
+                return ShuffledPlaylist.ToArray();
+            }
+            else
+            {
+                return SortedPlaylist.ToArray();
+            }
+        }
+
+        /// <summary>
+        /// Sets a new playlist. Playing item is set to none. Resets shuffle mode and repeat mode as well.
+        /// </summary>
+        /// <param name="items">The new items of the playlist.</param>
+        public void SetPlaylist(Guid[] items)
+        {
+            SortedPlaylist = CreateQueueItemsFromArray(items);
+            PlayingItemIndex = -1;
+            ShuffleMode = GroupShuffleMode.Sorted;
+            RepeatMode = GroupRepeatMode.RepeatNone;
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Appends new items to the playlist. The specified order is mantained for the sorted playlist, whereas items get shuffled for the shuffled playlist.
+        /// </summary>
+        /// <param name="items">The items to add to the playlist.</param>
+        public void Queue(Guid[] items)
+        {
+            var newItems = CreateQueueItemsFromArray(items);
+            SortedPlaylist.AddRange(newItems);
+
+            if (ShuffleMode.Equals(GroupShuffleMode.Shuffle))
+            {
+                newItems.Shuffle();
+                ShuffledPlaylist.AddRange(newItems);
```

Comment:
Wouldn't this better shuffle - as it would mix up the old and the new?

```suggestion
                ShuffledPlaylist.AddRange(newItems);
                ShuffledPlaylist.Shuffle();
```

Reply:
Right? This class is a translated copy of the client-side one. And lots of things didn't feel right to me either.
Shall we consider the shuffle request as a one time action or as a state?

I should probably review the class with a more server-side focused vision... :thinking: And bridge with the client at an upper level.

Reply:
Oh ok, I've checked again. So, in the original code items get pushed to the queue as they are, no shuffling is done.
My implementation diverged from the client-side one because the latter has a bug where items added to the queue are lost when returning to sorted mode.

So your proposed shuffle makes more sense, regardless of how it's implemented on the client.

Reply:
I've made my mind: the queue methods should not care about whether the playlist is shuffled or not, they just enqueue things and that's it. If the playlist needs to be re-shuffled then `ShufflePlaylist()` can be called again (the new implementation anyway).

What do you think?

Reply:
Might be worth adding that as a comment - shuffle is an action and not a state.